internal: Initial support for external client cert validation

tsaarni opened the pull request:
internal: Initial support for external client cert validation

Adds initial support for authentication of external clients (downstream) by
validating their client certificates against trusted CA certificate.

The feature is not yet exposed in the API.

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni commented:
This PR replaces #1226 by @uablrek.

I'm submitting this as a draft in order to ask for direction. @jpeach: in design doc review #1233 you mentioned 

> LGTM, with the proviso that in the final implementation, the upstream and downstream validation fields should use the same struct (i.e. the one currently named "UpstreamValidation").

In the `UpstreamValidation` we have validation of both CA and `subjectName` (or rather: Subject Alt Name).  The latter is not included in [design/tls-client-verification.md](design/tls-client-verification.md) since we wanted to minimize the feature as much possible.  I think I need two separate types in `apis/projectcontour/v1/httpproxy.go`: `UpstreamValidation` with `subjectName` and `DownstreamValidation` without?  

I have however combined these into single `ValidationContext` in `internal/dag/dag.go`.

Another alternative is to include `subjectName` also for downstream, but from usage perspective it would be rather a list of 0..N subjectNames, instead of single mandatory one like it is for upstream.

If we agree to keep two different types in `httpproxy.go`, I will then add similar test cases for  `DownstreamValidation` everywhere I find tests for  `UpstreamValidation`. For now the test coverage dropped because of this.

Obviously any other comments are of course welcome too!

jpeach commented:
Thanks @tsaarni; I'll review, but will need a few days to block out the time :)

davecheney commented:
Thank you all for your discussion. With my tech lead hat on I do need to remind everyone that the policy of this project is to talk, then code. We should not be designing features by duking it out in PR comments, that is not how we work on this project. 

Thank you

jpeach commented:
> On 20 Feb 2020, at 1:00 pm, Dave Cheney <notifications@github.com> wrote:
> 
> Thank you all for your discussion. With my tech lead hat on I do need to remind everyone that the policy of this project is to talk, then code. We should not be designing features by duking it out in PR comments, that is not how we work on this project.

Dave, this was previously discussed and a design doc was merged :)

https://github.com/projectcontour/contour/pull/1233
https://github.com/projectcontour/contour/blob/master/design/tls-client-verification.md

davecheney commented:
Oops, I am terribly sorry I mistook the discussion for the *other* auth discussions in fly at the moment. Please ignore me

jpeach commented:
> On 20 Feb 2020, at 1:07 pm, Dave Cheney <notifications@github.com> wrote:
> 
> Oops, I am terribly sorry I mistook the discussion for the other auth discussions in fly at the moment. Please ignore me

Yeh I can't find the PR for that, Github is making me terribly confused!

tsaarni commented:
I wonder if you would have time to provide any feedback for me? Thank you! :)

jpeach commented:
Sorry, I intended to review and then didn’t. I will review first thing next week.

> On Feb 28, 2020, at 6:15 PM, Tero Saarni <notifications@github.com> wrote:
> 
> I wonder if you would have time to provide any feedback for me? Thank you! :)

jpeach commented on internal/envoy/listener_test.go:
```diff
@@ -159,45 +160,104 @@ func TestSocketAddress(t *testing.T) {
+		"TLS context without client authentication": {
+			DownstreamTLSContext(secretName, envoy_api_v2_auth.TlsParameters_TLSv1_1, nil, "", "h2", "http/1.1"),
+			&envoy_api_v2_auth.DownstreamTlsContext{
+				CommonTlsContext: &envoy_api_v2_auth.CommonTlsContext{
```
You can hoist this into a variable to reduce copy/paste.

tsaarni replied:
Fixed.

jpeach commented on internal/dag/dag.go:
```diff
@@ -191,6 +191,14 @@ type UpstreamValidation struct {
+
+// ValidationContextCACert returns the CA certificate from ValidationContext
+func (vc *ValidationContext) ValidationContextCACert() []byte {
+	return vc.CACertificate.Object.Data[CACertificateKey]
```
If you check for `vc == nil`, then this function can be used to replace `upstreamValidationCACert`.

jpeach replied:
Done.

jpeach commented on internal/dag/dag.go:
```diff
@@ -191,6 +191,14 @@ type UpstreamValidation struct {
+const CACertificateKey = "ca.crt"
+
+// ValidationContextCACert returns the CA certificate from ValidationContext
+func (vc *ValidationContext) ValidationContextCACert() []byte {
```
This name stutters, just call it `GetCACertificate()`

jpeach replied:
Done.

jpeach commented on internal/dag/dag.go:
```diff
@@ -191,6 +191,14 @@ type UpstreamValidation struct {
+func (vc *ValidationContext) ValidationContextCACert() []byte {
+	return vc.CACertificate.Object.Data[CACertificateKey]
+}
+
```
Can you add the equivalent of `upstreamValidationSubjectAltName` as `GetSubjectName()`?

tsaarni replied:
Fixed.

jpeach commented on apis/projectcontour/v1/httpproxy.go:
```diff
@@ -119,6 +119,9 @@ type TLS struct {
 	// backing cluster.
 	// +optional
 	Passthrough bool `json:"passthrough,omitempty"`
+	// ClientValidation defines how to verify the client certificate
```
The audience for this comment are developers and operators trying to understand how to use the API, so we can err on the side of verbosity here. This setting

1. Enables TLS client certificate validation
1. Requires clients to present a TLS certificate (i.e. not optional validation)
1. Specifies how the client certificate will be validated

Do you think that covers it?

tsaarni replied:
Yes, that it good description.
I'm not sure if there should be some markup to get HTML list rendered correctly?

jpeach commented on internal/dag/builder.go:
```diff
@@ -487,6 +487,17 @@ func (b *Builder) computeHTTPProxy(proxy *projcontour.HTTPProxy) {
+			if tls.ClientValidation != nil {
+				dv, err := b.lookupDownstreamValidation(tls.ClientValidation,
+					proxy.Namespace,
+					fmt.Sprintf("HTTPProxy: %q", proxy.Name))
```
I made some minor cleanups to `lookupUpstreamValidation`, so here you should follow that pattern and add status message context in the caller.

tsaarni replied:
Fixed.

jpeach commented on internal/envoy/auth.go:
```diff
@@ -61,45 +62,50 @@ func UpstreamTLSContext(ca []byte, subjectName string, sni string, alpnProtocols
 }
 
-func validationContext(ca []byte, subjectName string) *envoy_api_v2_auth.CommonTlsContext_ValidationContext {
+func validationContext(ca []byte, subjectName string, subjectRequired bool) *envoy_api_v2_auth.CommonTlsContext_ValidationContext {
```
Do we really need the `subjectRequired` here? Since at the API level, the representation of upstream and downstream policy is disjoint, users should not be able to specify a subject name for downstream policy.

tsaarni replied:
I removed it. I needed to remove also the checks `len(ca)` and `len(subjectName)` from `validationContext()` and move these to `UpstreamTLSContext()`. I originally added `subjectRequired` since latter was not valid for `DownstreamTLSContext()`.

jpeach commented on internal/envoy/auth.go:
```diff
@@ -61,45 +62,50 @@ func UpstreamTLSContext(ca []byte, subjectName string, sni string, alpnProtocols
 // DownstreamTLSContext creates a new DownstreamTlsContext.
-func DownstreamTLSContext(secretName string, tlsMinProtoVersion envoy_api_v2_auth.TlsParameters_TlsProtocol, alpnProtos ...string) *envoy_api_v2_auth.DownstreamTlsContext {
-	return &envoy_api_v2_auth.DownstreamTlsContext{
+func DownstreamTLSContext(secretName string, tlsMinProtoVersion envoy_api_v2_auth.TlsParameters_TlsProtocol, ca []byte, subjectName string, alpnProtos ...string) *envoy_api_v2_auth.DownstreamTlsContext {
```
Here, and in `UpstreamTLSContext`, what would it look like if we passed a `*dag.PeerValidationPolicy` rather than `ca` and `subjectName`?

tsaarni replied:
Done. This triggered quite a lot of changes in unittests.

tsaarni commented on internal/dag/dag.go:
```diff
@@ -181,8 +181,8 @@ type HeaderValue struct {
-// UpstreamValidation defines how to validate the certificate on the upstream service
-type UpstreamValidation struct {
+// ValidationContext defines how to validate the certificate on the upstream service
+type ValidationContext struct {
```
I've renamed it to PeerValidationPolicy

tsaarni commented on internal/envoy/listener.go:
```diff
@@ -242,7 +242,7 @@ func FilterChains(filters ...*envoy_api_v2_listener.Filter) []*envoy_api_v2_list
 
 // FilterChainTLS returns a TLS enabled envoy_api_v2_listener.FilterChain,
-func FilterChainTLS(domain string, secret *dag.Secret, filters []*envoy_api_v2_listener.Filter, tlsMinProtoVersion envoy_api_v2_auth.TlsParameters_TlsProtocol, alpnProtos ...string) *envoy_api_v2_listener.FilterChain {
+func FilterChainTLS(domain string, secret *dag.Secret, clientCa []byte, clientSubjectName string, filters []*envoy_api_v2_listener.Filter, tlsMinProtoVersion envoy_api_v2_auth.TlsParameters_TlsProtocol, alpnProtos ...string) *envoy_api_v2_listener.FilterChain {
```
Now done. This one triggered quite a lot of changes in unittests!

tsaarni commented on internal/dag/dag.go:
```diff
@@ -191,6 +191,14 @@ type UpstreamValidation struct {
 }
 
+// CACertificateKey stores the key name for the TLS validation secret cert
+const CACertificateKey = "ca.crt"
```
Sure, I'll create separate PR!

jpeach replied:
See #2327

tsaarni commented on internal/contour/listener.go:
```diff
@@ -368,10 +368,12 @@ func (v *listenerVisitor) visit(vertex dag.Vertex) {
+				vh.Secret,
+				max(v.ListenerVisitorConfig.minProtoVersion(), vh.MinProtoVersion), // Choose the higher of the configured or requested tls version.
+				vh.DownstreamValidation,
+				alpnProtos...),
```
I've done this but I'm bit puzzled by this branch in general.  The logic was here before this PR, so I believe it is ok, but maybe I could add some comments here to make it easier to reason. 

- Here in `visit()` we have `dag.SecureVirtualHost`. I tend to interpret that as: we are going to terminate TLS
- We call `envoy.FilterChainTLS()` which again looks like we are going to terminate TLS
- However, when `vh.Secret` is `nil` then `envoy.FilterChainTLS()` will actually return filter chain without TLS termination?

I think I saw discussion between you and Dave on Slack that this case is for TCP proxy?  In that case `envoy.FilterChainTLS()` is not intended to construct actual TLS termination (unlike the function name may suggests) but just TCP passthrough *with SNI matching*?

jpeach replied:
Yes, if you look at the code on master, this always returns a filter chain match that matches the [SNI server name](https://www.envoyproxy.io/docs/envoy/v1.13.0/api-v3/config/listener/v3/listener_components.proto#envoy-v3-api-field-config-listener-v3-filterchainmatch-server-names), but it only configures downstream TLS if there is a secret.

tsaarni replied:
Thanks! I added small comment to clarify this:

```go
		// Secret is provided when TLS is terminated and nil when TLS passthrough is used.
		if vh.Secret != nil {
```

tsaarni commented on internal/envoy/auth.go:
```diff
@@ -48,69 +50,80 @@ var (
+func DownstreamTLSContext(serverSecret *dag.Secret, tlsMinProtoVersion envoy_api_v2_auth.TlsParameters_TlsProtocol, peerValidationContext *dag.PeerValidationContext, alpnProtos ...string) *envoy_api_v2_auth.DownstreamTlsContext {
+	if serverSecret == nil {
+		return nil
+	}
```
Done!

tsaarni commented on internal/envoy/cluster_test.go:
```diff
@@ -79,6 +79,17 @@ func TestCluster(t *testing.T) {
+				Name:      "secret",
+				Namespace: "default",
+			},
+			Type: "kubernetes.io/tls",
```
Done.
Maybe there should be another cleanup PR to change the rest since "kubernetes.io/tls" is all over?

jpeach replied:
I'd be happy to shepherd a PR that does that.

tsaarni commented on internal/featuretests/kubernetes.go:
```diff
@@ -90,6 +90,17 @@ func secretdata(cert, key string) map[string][]byte {
 	}
 }
 
+func caSecret(ca []byte) *v1.Secret {
```
Done!

tsaarni commented on apis/projectcontour/v1/httpproxy.go:
```diff
@@ -372,6 +378,12 @@ type UpstreamValidation struct {
+// DownstreamValidation defines how to verify the client certificate.
+type DownstreamValidation struct {
+	// Name of the Kubernetes secret be used to validate the certificate presented by the client
+	CACertificate string `json:"caSecret"`
```
Yep, thats right. Done!

tsaarni commented on internal/dag/builder.go:
```diff
@@ -488,6 +488,16 @@ func (b *Builder) computeHTTPProxy(proxy *projcontour.HTTPProxy) {
+			if tls.ClientValidation != nil {
+				dv, err := b.lookupDownstreamValidation(tls.ClientValidation, proxy.Namespace)
+				if err != nil {
+					sw.SetInvalid("HTTPProxy [%s]: TLS downstream validation policy error: %s", proxy.Name, err)
```
Done!

tsaarni commented on internal/envoy/listener_test.go:
```diff
@@ -158,46 +159,135 @@ func TestSocketAddress(t *testing.T) {
+				v1.TLSPrivateKeyKey: []byte("key"),
+			},
+		},
+	}
```
Done!

tsaarni commented on apis/projectcontour/v1/httpproxy.go:
```diff
@@ -372,6 +378,12 @@ type UpstreamValidation struct {
 
+// DownstreamValidation defines how to verify the client certificate.
+type DownstreamValidation struct {
+	// Name of the Kubernetes secret be used to validate the certificate presented by the client
```
Done!

jpeach commented:
@tsaarni The lint check failure should go away if you rebase to pick up the tooling script changes.

tsaarni commented:
I have prepared a commit that removes the API parts until #2347 is resolved.  In case you would still like to review this PR as "complete", I will not add that commit here in this PR until you confirm it is OK.

jpeach commented:
> I have prepared a commit that removes the API parts until #2347 is resolved. In case you would still like to review this PR as "complete", I will not add that commit here in this PR until you confirm it is OK.

Please go ahead and apply the commit to remove the API. Then, please squash commits, remove the "WIP" tag and update the PR subject to describe the current content of the commit. I'll commit the result.

tsaarni commented:
Thanks!  API removed, commits squashed, rebased and renamed!

Just as a reference in case needed, the API changes are stored on this unsquashed fork/branch https://github.com/Nordix/contour/tree/client-authentication-no-squash
